### PR TITLE
[navigation for multirotor] Sqrt Controller for Alt-Hold and improvements

### PR DESCRIPTION
**Pull request by JulioCesarMatias**
I don't know how much this will improve or worsen iNav Alt-Hold, figuring out the result of this implementation still needs testing. If this gives any results, it can also be used in Pos-Hold.

watch this video and skip to the 26:20 minute >> https://www.youtube.com/watch?v=-PC69jcMizA&ab_channel=ArduPilot

Tests:

- [x] Only sqrt controller (`Test Passed`)
- [x] sqrt controller + Feed Forward component (`Test Failed`)
- [x]  Change the Derivative LPF of PID Velocity Z to 5Hz (`Test Passed`)
- [x] Add LPF to PID Velocity Z controller error with CutOff freq in 5Hz (`Test Passed`)

The tests are being done by Jetrell.

### CODE REVIEW

File: src/main/navigation/navigation_multicopter.c [hunk 1]
```diff
@@ -72,6 +80,9 @@ static void updateAltitudeVelocityController_MC(timeDelta_t deltaMicros)
     else {
         targetVel = constrainf(targetVel, -navConfig()->general.max_auto_climb_rate, navConfig()->general.max_auto_climb_rate);
     }
+    
+    // add feed forward component
+    //targetVel += posControl.desiredState.vel.z;
```
Comment (JulioCesarMatias) [inline src/main/navigation/navigation_multicopter.c]: It is necessary to uncomment and test the Feed Forward component

File: src/main/navigation/navigation_multicopter.c [hunk 2]
```diff
@@ -105,7 +116,20 @@ static void updateAltitudeThrottleController_MC(timeDelta_t deltaMicros)
...
+    float accel_max_z = 0.0f;
+
+    if (posControl.flags.isAdjustingAltitude) {
+        accel_max_z = navConfig()->general.max_manual_climb_rate;
+    } else {
+        accel_max_z = navConfig()->general.max_auto_climb_rate;
+    }
+
+    // add feed forward component
+    //accel_target_z += constrainf(accel_desired_z, -accel_max_z, accel_max_z);
```
Comment (JulioCesarMatias) [inline src/main/navigation/navigation_multicopter.c:130]: It is necessary to uncomment and test the Feed Forward component

File: src/main/navigation/navigation_multicopter.c [hunk 3]
```diff
@@ -105,7 +116,22 @@ static void updateAltitudeThrottleController_MC(timeDelta_t deltaMicros)
     const int16_t thrAdjustmentMin = (int16_t)getThrottleIdleValue() - (int16_t)currentBatteryProfile->nav.mc.hover_throttle;
     const int16_t thrAdjustmentMax = (int16_t)motorConfig()->maxthrottle - (int16_t)currentBatteryProfile->nav.mc.hover_throttle;
 
-    posControl.rcAdjustment[THROTTLE] = navPidApply2(&posControl.pids.vel[Z], posControl.desiredState.vel.z, navGetCurrentActualPositionAndVelocity()->vel.z, US2S(deltaMicros), thrAdjustmentMin, thrAdjustmentMax, 0);
+    float accel_target_z = navPidApply2(&posControl.pids.vel[Z], posControl.desiredState.vel.z, navGetCurrentActualPositionAndVelocity()->vel.z, US2S(deltaMicros), thrAdjustmentMin, thrAdjustmentMax, 0);
+    
+/*
```
Comment (JulioCesarMatias) [inline src/main/navigation/navigation_multicopter.c]: It is necessary to uncomment and test the Feed Forward component

**Comment by JulioCesarMatias** [comment]
added the Feed Forward component in the Speed ​​and Acceleration control

**Comment by JulioCesarMatias** [comment]
first test sqrt_controller, and then test sqrt_controller + feed forward component

**Comment by Jetrell** [comment]
@JulioCesarMatias listen to the video explanation.... It sounds like a good improvement. Especial if it better predict MC slowdown from higher speeds to a specified stopping point, without the rubber band effect.

**Comment by JulioCesarMatias** [comment]
> 

ok, i was taking a look, and it seems to me that this has a chance of working well with iNav, i'm still waiting for the test results, the weather conditions here in brazil are not very good

File: src/main/navigation/navigation_multicopter.c (hunk 1 above)
Comment (JulioCesarMatias) [reply src/main/navigation/navigation_multicopter.c]: test failed

File: src/main/navigation/navigation_multicopter.c (hunk 3 above)
Comment (JulioCesarMatias) [reply src/main/navigation/navigation_multicopter.c]: test failed

**Comment by JulioCesarMatias** [comment]
@DzikuVx  this has been tested by @Jetrell several times and has given satisfactory results. What do you think of this PR? In case you find it interesting to merge, this square root controller can also be applied to Pos-Hold (I may open a PR in the future for this).

**Comment by Jetrell** [comment]
After testing this `sqrt controller`. I can conclude that it makes a significant difference to the way a MC slows down to a specified stopping point.  
It was tested under the following conditions -
* Flying in Acro, then switching to Poshold. 
* Flying in Poshold and releasing the stick.
* Auto navigation controlling RTH and stopping more precisely over the home landing location.
* Flying a Waypoint and stopping more precisely at the landing point.

All tests showed that the MC pulled up and stopped in a smoother and more accurate manor from speed, than with the present means.

Some of the tests were done at much higher speeds.  Even traveling at +120km/h and enabling Poshold or RTH. And in doing so, it showed no bad tendencies. _i.e. No instability or excessive rubber band pull-back / pendulum effect._

**Comment by JulioCesarMatias** [comment]
> After testing this `sqrt controller`. I can conclude that it makes a significant difference to the way a MC slows down to a specified stopping point.  
> 
> It was tested under the following conditions -
> 
> * Flying in Acro, then switching to Poshold. 
> 
> * Flying in Poshold and releasing the stick.
> 
> * Auto navigation controlling RTH and stopping more precisely over the home landing location.
> 
> * Flying a Waypoint and stopping more precisely at the landing point.
> 
> 
> 
> All tests showed that the MC pulled up and stopped in a smoother and more accurate manor from speed, than with the present means.
> 
> 
> 
> Some of the tests were done at much higher speeds.  Even traveling at +120km/h and enabling Poshold or RTH. And in doing so, it showed no bad tendencies. _i.e. No instability or excessive rubber band pull-back / pendulum effect._

Thank you so much for all the help you've given me. You are good friend! Maybe we can improve even more the AutoPilot of iNav, this square root controller can also be implemented for the correction of Position X and Y, in addition to Z that was made with this PR.

**Pull request closed without merging**